### ISSUE TITLE
Design Question, Is QMUIConfigurationTemplate necessary? 

### ISSUE THREAD

**Issue by caoer** [QMUI/QMUI_iOS#112]
### 运行环境 ###

- [x] iOS 设备：`iPhone` / `iPad` / `模拟器`
- [x] 系统版本：`iOS 10.x`
- [x] Xcode 版本：`8.x`
- [x] QMUI iOS 版本：`1.5.1`

### 具体问题描述 ###

Actually, it is not an issue, but rather a discussion. Would love to hear the feedback from QMUI team. 😄 

# QMUIConfigurationTemplate
Is QMUIConfigurationTemplate and calling [QMUIConfigurationManager renderGlobalAppearances] really necessary? Such design looks counter intuitive. QMUIKit is very high quality but in order to use it, one have to call two methods upon app start. I think it is unnecessary and should be redesigned. 

For following methods: 

```
  // 启动 QMUI 的样式配置模板
  [QMUIConfigurationTemplate setupConfigurationTemplate];
  // 将全局样式渲染出来
  [QMUIConfigurationManager renderGlobalAppearances];

```

`  [QMUIConfigurationTemplate setupConfigurationTemplate];` is calling `    [QMUICMI initDefaultConfiguration];` to setup default values. 

# QMUIConfigurationManager
`QMUICMI` is a macro defined as `#define QMUICMI [QMUIConfigurationManager sharedInstance]`. Here I feel strongly against this macro definition, the abbreviation makes code harder to be reasoned about. 

Let's take a look at `QMUIConfigurationManager`, it is a `configuration manager` class. However, it has a lot of properties attached to it. I don't think it makes sense because it is not doing any managing work. I purpose to change it to `QMUIConfiguration`. For example, Apple has `NSURLSessionConfiguration` which serves same functionality. it has a `defaultSessionConfiguration` class method which can return a default configuration. Later on, maybe we can serialize the 'Configuration' into JSON or plist so that it can be easily changed or load remotely. 

Another thing counter-intuitive about the QMUIConfigurationManager is that the default configuration is being set in the runtime. for defaultConfiguration, I think it is better to have them defined already. Library user doesn't have to initialize them before using it. For now, if I don't call `initDefaultConfiguration`. the app crashes. This makes it one step harder to use the library, and hard for other library developers to use it as a dependency. 

# renderGlobalAppearances

`renderGlobalAppearances` method looks ambiguous as the method name. It is a method not returning any object, so It doesn't render anything. What it does it setting up the appearance for each class. The UIApperance API is reasonable but a bit hard to work with. This approach makes it easier to reason about with writing less code. However, it introduces problems. We have to call those two methods in a right order, first setup the default configuration value, then we 'render' global appearance. If we changed the configuration, we shall call the 'render' method again to have it taking effects. 

I purpose to move them into the 'setter' method. for example, if we have only `QMUIConfigurationManager` class. it can have something like: 
```
- (void)setToolBarBarTintColor:(UIColor *)color {
    [UIToolbar appearance].barTintColor = color;
}

- (UIColor *)toolBarBarTintColor {
     return [UIToolbar appearance].barTintColor;
}
```
this way, the `manager` is not someone only talking but not doing things. He starts managing stuff. 😃 . Once I changed the color, it is taking effects. If someone changed the barTintColor, it is returning the correct value. 

How about have both `QMUIConfigurationManager` and `QMUIConfiguration`? This way, a configuration is a pure class not doing any work, the only purpose is storing configuration. The manager will be the one doing things. The values stored in configuration will for sure be that configuration. The value returned from manager will be the current value, even better!

That's it! this library is really good and did a lot of hardcore development. I've really appreciated it and hope it can become better and better.

**Comment by MoLice** [comment]
We really appreciate your suggestions and have released version [1.5.2](https://github.com/QMUI/QMUI_iOS/releases/tag/1.5.2) just now.

The following is a discussion about your advices.

>`QMUICMI` is a macro defined as `#define QMUICMI [QMUIConfigurationManager sharedInstance]`. Here I feel strongly against this macro definition, the abbreviation makes code harder to be reasoned about.

We’ve agreed it is not easy to understand. However, we define it mainly for reducing verbosity therefore the key codes can be highlighted. In general, the developers don't get in touch with it. So this macro remains unchanged in this updated version.

>I purpose to change it to QMUIConfiguration.

We have renamed `QMUIConfiguration` to `QMUIConfigurationMacros`, and `QMUIConfigurationManager` to `QMUIConfiguration`. In addition, default configuration will be set when `QMUIConfiguration` is initialized. In other words, `[QMUICMI initDefaultConfiguration]` in `QMUIConfigurationTemplate` has been removed.

>For now, if I don't call `initDefaultConfiguartion`. the app crashes.

This bug was caused because we didn’t work correctly when some properties were set to `nil`. And it has been fixed in a previous version [1.3.7](https://github.com/QMUI/QMUI_iOS/releases/tag/1.3.7) . Please update the version of QMUI.

>`renderGlobalAppearances` method looks ambiguous as the method name. ... I purpose to move them into the 'setter' method.

In this new version, the method above has been removed and the appearances will be updated in the 'setter' methods.
In other words, you just need one line to use QMUI in the new released version: 

```Objective-C
- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
    // 启动QMUI的配置模板
    [QMUIConfigurationTemplate setupConfigurationTemplate];
}
```

In the end, thank you again for your advises. We’re looking forward to your feedbacks of this newly released version.

**Comment by caoer** [comment]
I think this is not necessary to include QMUI? I actually don't need to write any code. Since the template is not doing anything by default. 
```
- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
    // 启动QMUI的配置模板
    [QMUIConfigurationTemplate setupConfigurationTemplate];
}
```

This newly version, 1.5.2, is much better and much easier to work with now, love it and appreciated the work being done so far. I'm working on a framework and going to include QMUIKit in it. 1.5.2 is good enough to be linked as a dependency. But for tech discussions only, I will state a few points, bare with me.  😄 

## Configuration, ConfigurationManager, and Singleton
Again, the current implementation is good enough, I will provide my input from my own experiences and thoughts to the road of a clean coder. 

Frist, singleton exists for a reason, take apple's design as an example. There aren't many singletons we can found in UIKit, but a few. UIApplication is singleton, it holds keyWindow and windows. which is reasonable for us, there is only one application we care about. Singleton should be used carefully, reasonably and responsibly, because it holds states, and when states go wrong, we have, bugs.... . Another example of the singleton is `[NSURLSesion sharedSession]`. I found the documentation of it really useful and explain a lot about the use cases and disadvantages of using sharedSession. Please head over to: https://developer.apple.com/reference/foundation/nsurlsession/1409000-sharedsession?language=objc . And at last, `[NSURLCache sharedURLCache]`, for URLCache, it is very reasonable to have it as singleton, since we don't want to have multiples inconsistent cache states. And please note that it is designed to let us change the sharedURLCache to provide customized behavior. 

Back to `QMUIConfiguration`, as of 1.5.2, in my own opinion, I think it might be better to call it, `QMUIConfigurationManager`. Because it is doing 'managing' work now, like setting up nav bar title font, background image, all sort of things. It also holds states, like clearColor, white color, yellowColor, etc. But change `QMUIConfiguration` into `QMUIConfigurationManager` doesn't make things any better. It only makes sense when we decoupled the states. In other words, they both need to exist. For example, `QMUIConfigurationManager` doesn't know what is blueColor, it relies on `QMUIConfiguration` to tell it. Once it knows what is blueColor, then it can apply it, for example,  `textFieldTintColor`. 

`QMUIConfiguration` now stores all kinds of appearance information, like color, edge insets, alpha, images, etc. If we can have multiple configurations, then we can have the ability to change app's appearance on the fly, or, maybe open a scheme gallery? any app using QMUIKit can have their users pick and choose a designed theme. A theme better for kids, teens, or visually impaired people. wow...

From current implementation, we can have multiple setup methods to achieve the switching appearance results, like `[QMUIConfigurationTemplate setupConfigurationTemplate];` and `[QMUIConfigurationTemplate setupVisuallyImpairedConfigurationTemplate];`. But in order to load settings from external resources, by passing setting values into `QMUIConfigurationTemplate`, then It looks like the same thing talked above.

**Comment by MoLice** [comment]
We just released a new version [1.6.0](https://github.com/QMUI/QMUI_iOS/releases/tag/1.6.0). In this version, we optimized the way to use QMUI especially the configuration template and `QMUIConfigurationTemplate` now became optional which means you can totally ignore it if you don't need it. Moreover, you can maintain more than one template just like our newest [QMUI Demo](https://github.com/QMUI/QMUIDemo_iOS) does.
Also, it is not necessary to call `[QMUIConfigurationTemplate setupConfigurationTemplate]` if you don't need it in this new version.
